apps/src/gamelab/spritelab/spriteUtils.js

```diff
@@ -0,0 +1,218 @@
+ * Returns a list of sprites, specified either by id or animation name.
+ * @param {(string|number)} spriteOrGroup - Either the id or the animation name
+ * @return {[Sprite]} List of sprites that match the parameter. Either a list containing the one sprite
+ * the specified id, or a list containing all sprites with the specified animation.
+ */
+export function singleOrGroup(spriteOrGroup) {
+  if (typeof spriteOrGroup === 'number') {
+    const sprite = nativeSpriteMap[spriteOrGroup];
+    if (sprite) {
+      return [sprite];
+    }
+  }
+  if (typeof spriteOrGroup === 'string') {
+    return allSpritesWithAnimation(spriteOrGroup);
+  }
+  return [];
+}
+
+export function getAnimationsInUse() {
+  let animations = new Set();
+  Object.keys(nativeSpriteMap).forEach(spriteId => {
+    animations.add(nativeSpriteMap[spriteId].getAnimationLabel());
+  });
+  return Array.from(animations);
+}
+
+export function getBehaviorsForAnimation(animation) {
+  let numBehaviors = 0;
+  behaviors.forEach(behavior => {
+    if (behavior.sprite.getAnimationLabel() === animation) {
+      numBehaviors++;
+    }
+  });
+  return numBehaviors;
+}
+
+export function getSpriteIdsInUse() {
+  let spriteIds = [];
+  Object.keys(nativeSpriteMap).forEach(spriteId =>
+    spriteIds.push(parseInt(spriteId))
+  );
+  return spriteIds;
+}
+
+/**
+ * Adds the specified sprite to the native sprite map
+ * @param {Sprite} sprite
+ * @returns {Number} A unique id to reference the sprite.
+ */
+export function addSprite(sprite) {
+  nativeSpriteMap[spriteId] = sprite;
+  sprite.id = spriteId;
+  spriteId++;
+  return sprite.id;
+}
+
+/**
+ * Removes a sprite from the native sprite map
+ * @param {Number} spriteId
+ */
+export function deleteSprite(spriteId) {
+  delete nativeSpriteMap[spriteId];
+}
+
+export function addEvent(type, args, callback) {
+  inputEvents.push({type: type, args: args, callback: callback});
+}
+
+function checkEvent(inputEvent, p5Inst) {
+  let shouldEventFire = false;
+  let extraArgs = {};
+  switch (inputEvent.type) {
+    case 'whenpress':
+      shouldEventFire = p5Inst.keyWentDown(inputEvent.args.key);
+      return {shouldEventFire: shouldEventFire, extraArgs: extraArgs};
+    case 'whilepress':
+      shouldEventFire = p5Inst.keyDown(inputEvent.args.key);
+      return {shouldEventFire: shouldEventFire, extraArgs: extraArgs};
+    case 'whentouch': {
+      let sprites = singleOrGroup(inputEvent.args.sprite1);
+      let targets = singleOrGroup(inputEvent.args.sprite2);
+      let overlap = false;
+      sprites.forEach(sprite => {
+        targets.forEach(target => {
+          if (sprite.overlap(target)) {
+            extraArgs.sprite = sprite.id;
+            extraArgs.target = target.id;
+            overlap = true;
+          }
+        });
+      });
+      if (overlap && !inputEvent.firedOnce) {
+        shouldEventFire = true;
+        inputEvent.firedOnce = true;
+      }
+      if (!overlap) {
+        inputEvent.firedOnce = false;
```

Comment:
This seems like it could introduce bugs if people start to rely on it for things besides overlap.

Reply:
Yeah, I'm not really sure how best to avoid that though. renamed to `firedOnceForCollision` so that at least it's not as surprising that it gets reset when sprites stop overlapping?

Reply:
Ahhhhh I get it now. I thought the `if(!overlap)` section was a no-op. But it needs to be able to collide multiple times and reset when it's no longer overlapping. Totally makes sense. Definitely the right way to do this. (just add a quick comment so it's a bit clearer)